src/main/drivers/bus_spi.c

```diff
@@ -128,116 +134,248 @@ bool spiInit(SPIDevice device, bool leadingEdge)
     return false;
 }
 
-uint32_t spiTimeoutUserCallback(SPI_TypeDef *instance)
+// Return true if DMA engine is busy
+bool spiIsBusy(const extDevice_t *dev)
 {
-    SPIDevice device = spiDeviceByInstance(instance);
-    if (device == SPIINVALID) {
-        return -1;
-    }
-    spiDevice[device].errorCount++;
-    return spiDevice[device].errorCount;
+    return (dev->bus->curSegment != (busSegment_t *)NULL);
 }
 
-bool spiBusTransfer(const busDevice_t *bus, const uint8_t *txData, uint8_t *rxData, int length)
+// Indicate that the bus on which this device resides may initiate DMA transfers from interrupt context
+void spiSetAtomicWait(const extDevice_t *dev)
 {
-    IOLo(bus->busdev_u.spi.csnPin);
-    spiTransfer(bus->busdev_u.spi.instance, txData, rxData, length);
-    IOHi(bus->busdev_u.spi.csnPin);
-    return true;
+    dev->bus->useAtomicWait = true;
 }
 
-uint16_t spiGetErrorCounter(SPI_TypeDef *instance)
+// Wait for DMA completion and claim the bus driver
+void spiWaitClaim(const extDevice_t *dev)
 {
-    SPIDevice device = spiDeviceByInstance(instance);
-    if (device == SPIINVALID) {
-        return 0;
+    // If there is a device on the bus whose driver might call spiSequence from an ISR then an
+    // atomic access is required to claim the bus, however if not, then interrupts need not be
+    // disabled as this can result in edge triggered interrupts being missed
+
+    if (dev->bus->useAtomicWait) {
+        // Prevent race condition where the bus appears free, but a gyro interrupt starts a transfer
+        do {
+            ATOMIC_BLOCK(NVIC_PRIO_MPU_INT_EXTI) {
```

Comment:
As noted via slack, the NVIC interrupt priority is specified as NVIC_PRIO_MPU_INT_EXTI
to me this doesn't make any sense because ANYTHING could be using the SPI to send data via an ISR.
The spi bus system shouldn't know anything about the gyro system.  Should the user of the spi bus (e.g. gyro code) be specifying the interrupt level?

The current codebase does not set `useAtomicWait` to true, PR #10573 does.  However, if you try and implement another driver that uses, say a timer ISR which has an NVIC priority of `NVIC_PRIO_TIMER`, to initiate SPI writes, then regardless of the `useAtomicWait` setting the code eventually ends up hanging in either of the two `while` loops in this method.

Reply:
We should therefore pass the required interrupt level to `useAtomicWait()`